### ISSUE TITLE
Consider expanding header rules to support company-defined policies

### ISSUE THREAD

**Issue by sharwell** [DotNetAnalyzers/StyleCopAnalyzers#1243]
Copyright headers are an aspect of source code which are often outside the ability of a local development team to influence. Unlike many other code style policies, switching to the StyleCop-defined file header structure may not be an option. It would be nice to consider expanding the ability of the file header rules to validate and automatically fix a wider variety of header policies. I'll start the discussion with the following header formats:

### Header 1

:memo: This form is used by my office. At the turn of each year we should see thousands of build warnings for existing code. :disappointed: 

```csharp
// Copyright (c) 2015 ACME, Inc. All rights reserved worldwide.
```

### Header 2

:memo: This form is used by corefx, Roslyn, and others.

```csharp
// Copyright (c) Microsoft. All rights reserved.
// Licensed under the MIT license. See LICENSE file in the project root for full license information.
```

### Header 3

:memo: This form was used in one of my older projects. I would probably switch it to use form 2 above if it was easy.

```
/*
 * Copyright (c) 2013 Sam Harwell, Tunnel Vision Laboratories LLC
 * All rights reserved.
 */
```

### Header 4

:memo: This form is used by two of my source-open (but not open-source) projects.

```csharp
/*
 *  Copyright (c) 2012 Sam Harwell, Tunnel Vision Laboratories LLC
 *  All rights reserved.
 *
 *  The source code of this document is proprietary work, and is not licensed for
 *  distribution. For information about licensing, contact Sam Harwell at:
 *      email@example.com
 */
```

**Comment by vweijsters** [comment]
I think it would be good to offer some sort of support for all of them.
The only reasonable way of implementing that I can think of at this moment is specifying the entire expected header text in the JSON file, with a special character (like `|`) to indicate line-breaks.

**Comment by sharwell** [comment]
> with a special character (like  `|`) to indicate line-breaks.

`\n` would probably be a good special character, since it's the JSON representation of an escaped newline character. :wink: The other possibility is encoding it using an array of strings, where each string in the array is one line.

**Comment by vweijsters** [comment]
I definitely need more JSON knowledge :smile: 
What kind of placeholders do you mean?

**Comment by sharwell** [comment]
For example, I would want the date to be a placeholder, so it knows the difference between a completely improperly formatted header and a header that's correct aside from the date. That way the code fix can update just the date.

**Comment by oatkins** [comment]
I have the year in my header, and update it only when something else in the file is changed. The legacy StyleCop rule is a nuisance, because it reports warnings for anything but the current year.

I would be glad to see any enhancement that handles this better.

**Comment by sharwell** [comment]
> I would be glad to see any enhancement that handles this better.

If it is within your control, I *highly* recommend you remove the year from the header. It makes everything easier.

**Comment by oatkins** [comment]
It certainly would make things easier. I'm not sure whether it *ought* to be present. Articles [like this one](http://www.contentious.com/2007/01/07/copyright-notice-is-the-year-really-necessary/), and the comments at the end, leave my head spinning.
